### PR TITLE
split `restore_training_state` into logical parts [2 / 2]

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #7652 and 
follow up to #7901 will be updated after #7901 merged!

Splits `restore_training_state` into the following pieces: 

```python
restore_callbacks()
restore_progress()
restore_optimizers()
restore_lr_schedulers()
```

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 1]
```diff
@@ -346,7 +379,9 @@ def hpc_load(self, checkpoint_path: str, on_gpu: bool):
             model.cuda(self.trainer.root_gpu)
 
         # restore training state
-        self.restore_training_state(checkpoint)
+        self._loaded_checkpoint = checkpoint
+        self.restore_training_state()
+        self._loaded_checkpoint = dict()
```
Comment (carmocca) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:384]: The fact that this is setting/unsetting is necessary makes me thing the checkpoint should be a method parameter

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:384]: good observation. had to put this here but it will disappear in the follow up pr, where hpc_load will get merged with the regular restore. 

hpc load does the same thing as the regular restore except call `on_hpc_load` instead of `on_load_checkpoint`. 
so I will be able to remove this code duplication. 

I can add a fixme comment if you agree?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py]: actually nevermind, reverted back so no need to set and reset this variable.

**Review by tchaton — APPROVED** [review]
Neat !

**Review by carmocca — APPROVED** [review]

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/checkpoint_connector.py [hunk 2]
```diff
@@ -135,77 +135,23 @@ def restore_model_state(self, model: LightningModule, checkpoint) -> None:
...
         """
-        Restore trainer state.
-        Model will get its change to update
-        :param checkpoint:
-        :return:
+        Restore the trainer state from the pre-loaded checkpoint. This includes the precision settings, loop progress,
+        optimizer states and learning rate scheduler states.
         """
+        if not checkpoint:
+            return
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/checkpoint_connector.py:144]: when can this happen?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:144]: when Trainer(resume_from_checkpoint=None) 
(the default case)
then trainer calls restore() and we do a no-op

Comment (ananthsub) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:144]: shouldn't we skip calling restore at all in that case?

Comment (awaelchli) [reply pytorch_lightning/trainer/connectors/checkpoint_connector.py:144]: no because it's actually an attempt to restore. 
this is the sequence:

1. attempt to restore from hpc_checkpoint file
2. attempt to restore from resume_from_checkpoint path if provided
3. do nothing

and could rename restore to attempt_restore if that's better?
